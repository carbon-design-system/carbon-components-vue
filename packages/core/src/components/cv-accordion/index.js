import CvAccordion from './cv-accordion';
import CvAccordionItem from './cv-accordion-item';
import CvAccordionSkeleton from './cv-accordion-skeleton';

export { CvAccordion, CvAccordionItem, CvAccordionSkeleton };
