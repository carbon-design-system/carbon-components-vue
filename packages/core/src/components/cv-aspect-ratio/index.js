import CvAspectRatio from './cv-aspect-ratio';
export { CvAspectRatio };
