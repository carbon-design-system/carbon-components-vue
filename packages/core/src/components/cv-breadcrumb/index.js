import CvBreadcrumb from './cv-breadcrumb';
import CvBreadcrumbItem from './cv-breadcrumb-item';
import CvBreadcrumbSkeleton from './cv-breadcrumb-skeleton';

export { CvBreadcrumb, CvBreadcrumbItem, CvBreadcrumbSkeleton };
