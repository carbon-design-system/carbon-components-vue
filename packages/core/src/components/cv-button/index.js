import CvButton from './cv-button';
import CvIconButton from './cv-icon-button';
import CvButtonSkeleton from './cv-button-skeleton';
import CvButtonSet from './cv-button-set';

export { CvButton, CvIconButton, CvButtonSkeleton, CvButtonSet };
