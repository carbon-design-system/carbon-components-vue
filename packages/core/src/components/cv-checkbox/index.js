import CvCheckbox from './cv-checkbox';
import CvCheckboxSkeleton from './cv-checkbox-skeleton';
export { CvCheckbox, CvCheckboxSkeleton };
