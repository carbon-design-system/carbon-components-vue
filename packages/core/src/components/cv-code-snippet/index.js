import CvCodeSnippet from './cv-code-snippet';
import CvCodeSnippetSkeleton from './cv-code-snippet-skeleton';
export { CvCodeSnippet, CvCodeSnippetSkeleton };
