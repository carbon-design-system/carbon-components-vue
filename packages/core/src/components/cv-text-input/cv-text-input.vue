<template>
  <div :class="`cv-text-input ${carbonPrefix}--form-item`">
    <label
      :for="uid"
      :class="[
        `${carbonPrefix}--label`,
        {
          [`${carbonPrefix}--label--disabled`]: $attrs.disabled !== undefined && $attrs.disabled,
        },
      ]"
      >{{ label }}</label
    >
    <div :class="`${carbonPrefix}--text-input__field-wrapper`" :data-invalid="isInvalid">
      <WarningFilled16 v-if="isInvalid" :class="`${carbonPrefix}--text-input__invalid-icon`" />
      <input
        :id="uid"
        :class="[
          `${carbonPrefix}--text-input`,
          {
            [`${carbonPrefix}--text-input--light`]: theme === 'light',
            [`${carbonPrefix}--text-input--invalid`]: isInvalid,
          },
        ]"
        v-bind="$attrs"
        :value="value"
        v-on="inputListeners"
        :data-toggle-password-visibility="isPassword"
        :type="dataType"
        ref="input"
      />
      <button
        v-if="isPassword"
        :class="[
          `${carbonPrefix}--text-input--password__visibility__toggle`,
          `${carbonPrefix}--tooltip__trigger`,
          `${carbonPrefix}--tooltip--a11y`,
          `${carbonPrefix}--tooltip--bottom`,
          `${carbonPrefix}--tooltip--align-center`,
        ]"
        @click="togglePasswordVisibility"
        type="button"
      >
        <span :class="`${carbonPrefix}--assistive-text`">{{ passwordHideShowLabel }}</span>
        <ViewOff16 v-if="isPasswordVisible" :class="`${carbonPrefix}--icon-visibility-off`" />
        <View16 v-else :class="`${carbonPrefix}--icon-visibility-off`" />
      </button>
    </div>
    <div :class="`${carbonPrefix}--form-requirement`" v-if="isInvalid">
      <slot name="invalid-message">{{ invalidMessage }}</slot>
    </div>
    <div
      v-if="!isInvalid && isHelper"
      :class="[
        `${carbonPrefix}--form__helper-text`,
        { [`${carbonPrefix}--form__helper-text--disabled`]: $attrs.disabled },
      ]"
    >
      <slot name="helper-text">{{ helperText }}</slot>
    </div>
  </div>
</template>

<script>
import { uidMixin, themeMixin, methodsMixin, carbonPrefixMixin } from '../../mixins';
import WarningFilled16 from '@carbon/icons-vue/es/warning--filled/16';
import View16 from '@carbon/icons-vue/es/view/16';
import ViewOff16 from '@carbon/icons-vue/es/view--off/16';

export default {
  name: 'CvTextInput',
  components: { WarningFilled16, View16, ViewOff16 },
  mixins: [uidMixin, themeMixin, carbonPrefixMixin, methodsMixin({ input: ['blur', 'focus'] })],
  inheritAttrs: false,
  props: {
    helperText: { type: String, default: undefined },
    invalidMessage: { type: String, default: undefined },
    label: String,
    passwordHideLabel: { type: String, default: 'Hide password' },
    passwordShowLabel: { type: String, default: 'Show password' },
    passwordVisible: Boolean,
    theme: String,
    type: String,
    value: String,
  },
  data() {
    return {
      dataPasswordVisible: this.isPassword && this.passwordVisible,
      dataType: this.type,
      isHelper: false,
      isInvalid: false,
    };
  },
  mounted() {
    this.checkSlots();
  },
  updated() {
    this.checkSlots();
  },
  watch: {
    passwordVisible() {
      if (this.passwordVisible !== this.dataPasswordVisible) {
        this.togglePasswordVisibility();
      }
    },
    type() {
      this.dataType = this.type;
    },
  },
  computed: {
    // Bind listeners at the component level to the embedded input element and
    // add our own input listener to service the v-model. See:
    // https://vuejs.org/v2/guide/components-custom-events.html#Customizing-Component-v-model
    inputListeners() {
      return {
        ...this.$listeners,
        input: event => this.$emit('input', event.target.value),
      };
    },
    isPassword() {
      return this.type === 'password';
    },
    isPasswordVisible() {
      return this.isPassword && this.dataPasswordVisible;
    },
    passwordHideShowLabel() {
      return this.isPasswordVisible ? this.passwordHideLabel : this.passwordShowLabel;
    },
  },
  methods: {
    checkSlots() {
      // NOTE: this.$slots is not reactive so needs to be managed on updated
      this.isInvalid = !!(this.$slots['invalid-message'] || (this.invalidMessage && this.invalidMessage.length));
      this.isHelper = !!(this.$slots['helper-text'] || (this.helperText && this.helperText.length));
    },
    togglePasswordVisibility() {
      const currentValue = this.$refs.input.value;
      this.dataPasswordVisible = !this.dataPasswordVisible;
      this.dataType = this.dataPasswordVisible ? 'text' : 'password';

      this.$nextTick(() => {
        this.$refs.input.value = currentValue;
      });
    },
  },
};
</script>
