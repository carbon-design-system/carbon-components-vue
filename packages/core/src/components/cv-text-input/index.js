import CvTextInput from './cv-text-input';
export { CvTextInput };
