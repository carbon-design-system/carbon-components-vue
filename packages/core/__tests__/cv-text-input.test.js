// import { shallowMount as shallow } from '@vue/test-utils';
import { testComponent, awaitNextTick } from './_helpers';
const { shallowMount: shallow, trigger, setProps } = awaitNextTick;
import { CvTextInput } from '@/components/cv-text-input';

describe('CvTextInput', () => {
  const TYPE_PASSWORD = 'password';
  const TYPE_TEXT = 'text';
  // ***************
  // PROP CHECKS
  // ***************
  testComponent.propsAreType(
    CvTextInput,
    ['helperText', 'invalidMessage', 'label', 'passwordHideLabel', 'passwordShowLabel', 'type', 'value', 'id'],
    String
  );
  testComponent.propsAreType(CvTextInput, ['passwordVisible'], Boolean);
  testComponent.propsHaveDefaultOfUndefined(CvTextInput, ['helperText', 'invalidMessage', 'light']);
  testComponent.propsHaveDefault(CvTextInput, ['passwordHideLabel', 'passwordShowLabel']);

  // ***************
  // SNAPSHOT TESTS
  // ***************

  it('should render correctly', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1', type: TYPE_TEXT };
    const wrapper = await shallow(CvTextInput, { propsData });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when light theme is used', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1', theme: 'light', type: TYPE_TEXT };
    const wrapper = await shallow(CvTextInput, { propsData });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when invalid message is provided', async () => {
    const propsData = {
      label: 'test label',
      value: 'test value',
      invalidMessage: 'invalid test message',
      id: '1',
      type: TYPE_TEXT,
    };
    const wrapper = await shallow(CvTextInput, { propsData });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when invalid message slot is provided', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1', type: TYPE_TEXT };
    const wrapper = await shallow(CvTextInput, {
      slots: {
        'invalid-message': '<div class="invalid-message-class">invalid message slot</div>',
      },
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when helper text is provided', async () => {
    const propsData = {
      label: 'test label',
      value: 'test value',
      helperText: 'helper test message',
      id: '1',
      type: TYPE_TEXT,
    };
    const wrapper = await shallow(CvTextInput, { propsData });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when helper text slot is provided', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1', type: TYPE_TEXT };
    const wrapper = await shallow(CvTextInput, {
      slots: {
        'helper-text': '<div class="helper-text-class">helper text slot</div>',
      },
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when type is password and is visible', async () => {
    const propsData = {
      label: 'test label',
      value: 'test password value',
      id: '1',
      type: TYPE_PASSWORD,
      passwordVisible: true,
    };
    const wrapper = await shallow(CvTextInput, { propsData });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should render correctly when type is password and is not visible', async () => {
    const propsData = {
      label: 'test label',
      value: 'test password value',
      id: '1',
      type: TYPE_PASSWORD,
      passwordVisible: false,
    };
    const wrapper = await shallow(CvTextInput, { propsData });
    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************

  it('`passwordVisible` prop should be watched', async () => {
    const propsData = { value: 'test value', id: '1', type: TYPE_PASSWORD, passwordVisible: false };
    const wrapper = await shallow(CvTextInput, { propsData });
    const spy = jest.spyOn(wrapper.vm, 'togglePasswordVisibility');
    await setProps(wrapper, {
      passwordVisible: true,
    });
    expect(spy).toBeCalledTimes(1);
  });

  it('`type` prop should be watched', async () => {
    const propsData = { value: 'test value', id: '1', type: TYPE_PASSWORD };
    const wrapper = await shallow(CvTextInput, { propsData });
    expect(wrapper.vm.dataType).toEqual(TYPE_PASSWORD);
    await setProps(wrapper, {
      type: TYPE_TEXT,
    });
    expect(wrapper.vm.dataType).toEqual(TYPE_TEXT);
  });

  it('should emit current value on change', async () => {
    const value = 'test value';
    const propsData = { value, id: '1' };
    const wrapper = await shallow(CvTextInput, { propsData });
    await trigger(wrapper.find('input'), 'input');
    expect(wrapper.emitted().input[0]).toEqual([value]);
  });

  it('`isPassword` should be true when type is `password`', async () => {
    const propsData = { value: 'test value', id: '1', type: TYPE_PASSWORD };
    const wrapper = await shallow(CvTextInput, { propsData });
    expect(wrapper.vm.isPassword).toEqual(true);
  });

  it('`isPassword` should be false when type is not `password`', async () => {
    const propsData = { value: 'test value', id: '1', type: TYPE_TEXT };
    const wrapper = await shallow(CvTextInput, { propsData });
    expect(wrapper.vm.isPassword).toEqual(false);
  });

  it('`isPasswordVisible` should be true when type is `password` and password is visible', async () => {
    const propsData = { value: 'test value', id: '1', type: TYPE_PASSWORD, passwordVisible: true };
    const wrapper = await shallow(CvTextInput, { propsData });
    wrapper.setData({
      dataPasswordVisible: true,
    });
    expect(wrapper.vm.isPasswordVisible).toEqual(true);
  });

  it('`isPasswordVisible` should be false when type is `password` but password is not visible', async () => {
    const propsData = { value: 'test value', id: '1', type: TYPE_PASSWORD, passwordVisible: false };
    const wrapper = await shallow(CvTextInput, { propsData });
    wrapper.setData({
      dataPasswordVisible: false,
    });
    expect(wrapper.vm.isPasswordVisible).toEqual(false);
  });

  it('`isPasswordVisible` should be false when type is not `password` even if `passwordVisible` prop is set to true', async () => {
    const propsData = { value: 'test value', id: '1', type: TYPE_TEXT, passwordVisible: true };
    const wrapper = await shallow(CvTextInput, { propsData });
    wrapper.setData({
      dataPasswordVisible: false,
    });
    expect(wrapper.vm.isPasswordVisible).toEqual(false);
  });

  it('`passwordHideShowLabel` should be equal to `passwordHideLabel` prop when password is visible', async () => {
    const passwordHideLabel = ' password hide label test';
    const propsData = { value: 'test value', id: '1', type: TYPE_PASSWORD, passwordVisible: true, passwordHideLabel };
    const wrapper = await shallow(CvTextInput, { propsData });
    wrapper.setData({
      dataPasswordVisible: true,
    });
    expect(wrapper.vm.passwordHideShowLabel).toEqual(passwordHideLabel);
  });

  it('`passwordHideShowLabel` should be equal to `passwordShowLabel` prop when password is visible', async () => {
    const passwordShowLabel = ' password show label test';
    const propsData = { value: 'test value', id: '1', type: TYPE_PASSWORD, passwordVisible: false, passwordShowLabel };
    const wrapper = await shallow(CvTextInput, { propsData });
    expect(wrapper.vm.passwordHideShowLabel).toEqual(passwordShowLabel);
  });

  it('`togglePasswordVisibility` works as expected', async () => {
    const propsData = { value: 'test value', id: '1', type: TYPE_PASSWORD, passwordVisible: false };
    const wrapper = await shallow(CvTextInput, { propsData });
    expect(wrapper.vm.dataType).toEqual(TYPE_PASSWORD);
    wrapper.vm.togglePasswordVisibility();
    expect(wrapper.vm.dataType).toEqual(TYPE_TEXT);
  });
});
