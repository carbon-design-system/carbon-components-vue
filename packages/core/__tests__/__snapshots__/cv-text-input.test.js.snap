// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvTextInput should render correctly 1`] = `
<div class="cv-text-input bx--form-item bx--text-input-wrapper"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-input__field-wrapper">
    <!---->
    <!----> <input id="1" type="text" class="bx--text-input">
    <!---->
  </div>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`CvTextInput should render correctly when helper text is provided 1`] = `
<div class="cv-text-input bx--form-item bx--text-input-wrapper"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-input__field-wrapper">
    <!---->
    <!----> <input id="1" type="text" class="bx--text-input">
    <!---->
  </div>
  <!---->
  <!---->
  <div class="bx--form__helper-text">helper test message</div>
</div>
`;

exports[`CvTextInput should render correctly when helper text slot is provided 1`] = `
<div class="cv-text-input bx--form-item bx--text-input-wrapper"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-input__field-wrapper">
    <!---->
    <!----> <input id="1" type="text" class="bx--text-input">
    <!---->
  </div>
  <!---->
  <!---->
  <div class="bx--form__helper-text">
    <div class="helper-text-class">helper text slot</div>
  </div>
</div>
`;

exports[`CvTextInput should render correctly when invalid message is provided 1`] = `
<div class="cv-text-input bx--form-item bx--text-input-wrapper"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-input__field-wrapper" data-invalid="true">
    <warningfilled16-stub class="bx--text-input__invalid-icon"></warningfilled16-stub>
    <!----> <input id="1" type="text" class="bx--text-input bx--text-input--invalid">
    <!---->
  </div>
  <div class="bx--form-requirement">invalid test message</div>
  <!---->
  <!---->
</div>
`;

exports[`CvTextInput should render correctly when invalid message slot is provided 1`] = `
<div class="cv-text-input bx--form-item bx--text-input-wrapper"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-input__field-wrapper" data-invalid="true">
    <warningfilled16-stub class="bx--text-input__invalid-icon"></warningfilled16-stub>
    <!----> <input id="1" type="text" class="bx--text-input bx--text-input--invalid">
    <!---->
  </div>
  <div class="bx--form-requirement">
    <div class="invalid-message-class">invalid message slot</div>
  </div>
  <!---->
  <!---->
</div>
`;

exports[`CvTextInput should render correctly when light theme is used 1`] = `
<div class="cv-text-input bx--form-item bx--text-input-wrapper"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-input__field-wrapper">
    <!---->
    <!----> <input id="1" type="text" class="bx--text-input bx--text-input--light">
    <!---->
  </div>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`CvTextInput should render correctly when type is password and is not visible 1`] = `
<div class="cv-text-input bx--form-item bx--text-input-wrapper bx--password-input-wrapper"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-input__field-wrapper">
    <!---->
    <!----> <input id="1" data-toggle-password-visibility="true" type="password" class="bx--text-input bx--password-input"> <button type="button" class="bx--btn bx--text-input--password__visibility__toggle bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--bottom bx--tooltip--align-center"><span class="bx--assistive-text">Show password</span>
      <view16-stub class="bx--icon-visibility-off"></view16-stub>
    </button></div>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`CvTextInput should render correctly when type is password and is visible 1`] = `
<div class="cv-text-input bx--form-item bx--text-input-wrapper bx--password-input-wrapper"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-input__field-wrapper">
    <!---->
    <!----> <input id="1" data-toggle-password-visibility="true" type="password" class="bx--text-input bx--password-input"> <button type="button" class="bx--btn bx--text-input--password__visibility__toggle bx--tooltip__trigger bx--tooltip--a11y bx--tooltip--bottom bx--tooltip--align-center"><span class="bx--assistive-text">Show password</span>
      <view16-stub class="bx--icon-visibility-off"></view16-stub>
    </button></div>
  <!---->
  <!---->
  <!---->
</div>
`;
